src/common/base/CollectNSucceeded-inl.h

```diff
@@ -11,59 +11,109 @@ namespace nebula {
   // we have required number of futures, at which point we fulfil
   // the promise with the result list
+  auto ctx = std::make_shared<Context>(total, std::forward<ResultEval>(eval));
   for (size_t index = 0; first != last; ++first, ++index) {
-    first->setCallback_([n, ctx, index](auto, folly::Try<FutureReturnType<FutureIter>>&& t) {
-      if (!ctx->promise.isFulfilled()) {
+    first->setCallback_([n, ctx, index](auto&&,
+                                        folly::Try<FutureReturnType<FutureIter>>&& t) {
+      VLOG(3) << "Received the callback from the future " << index;
+
+      // relaxed because this guards control but does not guard data
+      auto const c = 1 + ctx->nCompleted.fetch_add(1, std::memory_order_relaxed);
+      auto s = ctx->nSucceeded.load();
+
+      if (!ctx->fulfilled.load()) {
         if (!t.hasException()) {
           if (ctx->eval(index, t.value())) {
-            ++ctx->nSucceeded;
+            // release because the stored values in all threads must be visible below
+            // acquire because no stored value is permitted to be fetched early
+            s = 1 + ctx->nSucceeded.fetch_add(1, std::memory_order_acq_rel);
           }
-          ctx->results.emplace_back(index, std::move(t.value()));
+        } else {
+          // Got exception
+          DLOG(ERROR) << "Exception receided: " << t.exception().what().toStdString();
+        }
+
+        ctx->results[index] = std::move(t);
+        if (c < ctx->nTotal && s < n) {
+          return;
         }
-        if ((++ctx->numCompleted) == ctx->nTotal || ctx->nSucceeded == n) {
-          // Done
-          VLOG(2) << "Set Value [completed=" << ctx->numCompleted << ", total=" << ctx->nTotal
-                  << ", Result list size=" << ctx->results.size() << "]";
-          ctx->promise.setValue(std::move(ctx->results));
+
+        bool fulfilled = false;
+        if (!ctx->fulfilled.compare_exchange_strong(fulfilled, true)) {
+          // Already fulfilled
+          return;
+        }
+
+        Result collectedResults;
+        collectedResults.reserve(ctx->nCompleted.load());
+        for (size_t i = 0; i < ctx->nTotal; ++i) {
+          auto& entry = ctx->results[i];
+          if (entry.hasValue() && entry.value().hasValue()) {
+            collectedResults.emplace_back(i, std::move(entry).value().value());
+          }
         }
+
+        // Done
+        VLOG(2) << "Set Value [total=" << ctx->nTotal
+                << ", completed=" << c
+                << ", succeeded=" << s
+                << ", Result list size=" << collectedResults.size() << "]";
+        ctx->promise.setValue(std::move(collectedResults));
       }
     });
+    VLOG(3) << "Set up future callback " << index;
   }
 
-  return ctx->promise.getFuture();
+  auto future = ctx->promise.getSemiFuture();
+  if (!executors.empty()) {
+    VLOG(3) << "Wiring the nested executors";
+    future = std::move(future).defer(
```

Comment:
Any special reason to defer work on consumer side? I notice that you changed `result` to `std::vector<folly::Optional<folly::Try<FutureReturnType<FutureIter>>>>`.